crates/router/src/core/user_role/role.rs

```diff
@@ -322,3 +322,147 @@ pub async fn update_role(
+        })
+    });
+
+    Ok(ApplicationResponse::Json(
+        predefined_roles_map.chain(custom_roles_map).collect(),
+    ))
+}
+
+pub async fn list_roles_at_entity_level(
+    state: SessionState,
+    user_from_token: UserFromToken,
+    req: role_api::ListRolesAtEntityLevelRequest,
+    check_type: role_api::RoleCheckType,
+) -> UserResponse<Vec<role_api::MinimalRoleInfo>> {
+    let user_entity_type = user_from_token
+        .get_role_info_from_db(&state)
+        .await
+        .attach_printable("Invalid role_id in JWT")?
+        .get_entity_type();
+
+    if req.entity_type > user_entity_type {
+        return Err(UserErrors::InvalidRoleOperationWithMessage(
+            "User is attempting to request list roles above the current entity level".to_string(),
+        )
+        .into());
+    }
+
+    let predefined_roles_map = PREDEFINED_ROLES
+        .iter()
+        .filter(|(_, role_info)| {
+            let check_type = match check_type {
+                role_api::RoleCheckType::Invite => role_info.is_invitable(),
+                role_api::RoleCheckType::Update => role_info.is_updatable(),
+            };
+            check_type && role_info.get_entity_type() == req.entity_type
+        })
+        .map(|(role_id, role_info)| role_api::MinimalRoleInfo {
+            role_id: role_id.to_string(),
+            role_name: role_info.get_role_name().to_string(),
+        });
+
+    let custom_roles = match req.entity_type {
+        EntityType::Organization => state
+            .store
+            .list_roles_for_org_by_parameters(
+                &user_from_token.org_id,
+                None,
+                Some(req.entity_type),
+                None,
+            )
+            .await
+            .change_context(UserErrors::InternalServerError)
+            .attach_printable("Failed to get roles")?,
+
+        EntityType::Merchant => state
+            .store
+            .list_roles_for_org_by_parameters(
+                &user_from_token.org_id,
+                Some(&user_from_token.merchant_id),
+                Some(req.entity_type),
+                None,
+            )
+            .await
+            .change_context(UserErrors::InternalServerError)
+            .attach_printable("Failed to get roles")?,
+        // TODO: Populate this from Db function when support for profile id and profile level custom roles is added
+        EntityType::Profile => Vec::new(),
+        EntityType::Internal => {
+            return Err(UserErrors::InvalidRoleOperationWithMessage(
+                "Internal roles are not allowed for this operation".to_string(),
+            )
+            .into());
+        }
+    };
+
+    let custom_roles_map = custom_roles.into_iter().map(|role| {
+        let role_info = roles::RoleInfo::from(role);
+        role_api::MinimalRoleInfo {
+            role_id: role_info.get_role_id().to_string(),
+            role_name: role_info.get_role_name().to_string(),
+        }
```

Comment:
Add entity type check and is_updatable / is_invitable.

Reply:
entity type is already handled in db query, added check for invitable/updatable